cxx/distributions/skellam.cc

```diff
@@ -0,0 +1,66 @@
+#include "distributions/skellam.hh"
+
+#include <cassert>
+#include <cmath>
+#include "util_math.hh"
+
+double lognormal_logp(double x, double mean, double stddev) {
+  double y = (std::log(x) - mean) / stddev;
+  return - y*y / 2.0
+      - std::log(x * stddev) - 0.5 * std::log(2.0 * std::numbers::pi);
+}
+
+double Skellam::logp(const int&x) const {
+  return -mu1 - mu2 + (x / 2.0) * std::log(mu1 / mu2)
+      // TODO(thomaswc): Replace this with something more numerically stable.
+      + std::log(std::cyl_bessel_i(x, 2.0 * std::sqrt(mu1 * mu2)));
+}
+
+int Skellam::sample(std::mt19937* prng) {
+  std::poisson_distribution<int> d1(mu1);
+  std::poisson_distribution<int> d2(mu2);
+  return d1(*prng) - d2(*prng);
+}
+
+void Skellam::transition_hyperparameters(std::mt19937* prng) {
+  std::vector<double> logps;
+  std::vector<std::tuple<double, double, double, double>> hypers;
+  for (double tmean1 : MEAN_GRID) {
+    for (double tstddev1 : STDDEV_GRID) {
+      for (double tmean2 : MEAN_GRID) {
+        for (double tstddev2 : STDDEV_GRID) {
+          double lp = lognormal_logp(mu1, tmean1, tstddev1)
+              + lognormal_logp(mu2, tmean2, tstddev2);
+          logps.push_back(lp);
+          hypers.push_back(
+              std::make_tuple(tmean1, tstddev1, tmean2, tstddev2));
+        }
+      }
+    }
+  }
+  int i = sample_from_logps(logps, prng);
+  mean1 = std::get<0>(hypers[i]);
+  stddev1 = std::get<1>(hypers[i]);
+  mean2 = std::get<2>(hypers[i]);
+  stddev2 = std::get<3>(hypers[i]);
+}
+
+void Skellam::init_theta(std::mt19937* prng) {
```

Comment:
Out of curiosity, is there a need for Skellam to be initialized randomly vs. statically like the other distributions? Or do we see other distributions initializing their parameters from their hyperprior? Mainly asking since it seems a little odd that Skellam's path is different from other distributions.

Reply:
Yeah, all the NonconjugateDistributions will need random initializations.  The reason is that the NonconjugateDistributions can't efficiently marginalize over the latent parameters, so instead they store current values of their latents, which evolve whenever transition_theta is called.  And I believe that that overall learning procedure works better when the latents are randomly initialized, but I guess I could be wrong about that.  I personally always use random initialization when doing Metropolis-Hastings, but I guess some people always use the origin or the mean of the sampling distribution or something like that.

Anyway, we could definitely raise this issue on the slack channel if you like, but the above is my current understanding.  Oh, and I guess another thing is that the GenDB doc says that for "distributions that explicitly represent their latents", "the code to initialize the model state will need to include code for sampling initial parameters from the parameter prior".  So this is that.